Casmcms 8742 1.6 (#4910)
Cray-HPE/docs-csm: operations/image_management/Troubleshoot_Remote_Build_Node.md

Reviewer: Need to add links to this file to the main operations index (`operations/README.md`), the troubleshooting index (`troubleshooting/README.md`), and any relevant IMS pages in the `operations/image_management` folder.

# Troubleshoot Remote Build Node

> Remote builds work by creating and running containerized jobs via `podman` on a remote node. There

Reviewer: ```suggestion
Remote builds work by creating and running containerized jobs via podman on a remote node. There
```

are times that problems crop up with running these remote jobs.

* [Prerequisites](#prerequisites)
* [Jobs fail to launch on remote nodes](#jobs-fail-to-launch-on-remote-nodes)
* [Clean up orphaned jobs](#clean-up-orphaned-jobs)
* [Jobs fail due to lack of resources on the remote node](#jobs-fail-due-to-lack-of-resources-on-the-remote-node)

## Prerequisites

* Administrative access to a master or worker node on the K8S cluster
* Administrative access to the remote build node

## Jobs fail to launch on remote nodes

1. Verify that the remote node is defined in IMS.

    For more information, see [Adding and removing remote build nodes to IMS](Configure_a_Remote_Build_Node.md#adding-and-removing-remote-build-nodes-to-ims).

1. Verify that SSH keys are installed on the remote node.

    For more information, see [Use an existing compute node](Configure_a_Remote_Build_Node.md#use-an-existing-compute-node#).

> 1. Verify that `podman` is installed on the remote node.

Reviewer: ```suggestion
1. Verify that podman is installed on the remote node.
```

    For more information, see [Use an existing compute node](Configure_a_Remote_Build_Node.md#use-an-existing-compute-node#).

1. Check the architecture of of the job and remote node.

    1. (`ncn-mw#`) Check job architecture.

        ```bash

>         cray ims jobs describe JOB_ID

Reviewer: ```suggestion
        cray ims jobs describe JOB_ID --format json
```

    [2024-04-19 19:03:45,342] INFO in helper: ++ _get_s3_download_url {'etag': 'e801003a78039843210cc02f8be1248e-192', 'path': 's3://boot-images/f65534d9-c469-4bab-8bb3-f66c3cf98042/rootfs', 'type': 's3'}.
    [2024-04-19 19:03:45,343] INFO in jobs: ARTIFACT_RECORD: <V2ImageRecord(id=UUID('f65534d9-c469-4bab-8bb3-f66c3cf98042'))>
    [2024-04-19 19:03:45,343] INFO in jobs: architecture: aarch64
    [2024-04-19 19:03:45,344] INFO in jobs: kernel file name: Image
    [2024-04-19 19:03:45,344] INFO in jobs:  NOTE: aarch64 architecture requires dkms
    [2024-04-19 19:03:45,344] INFO in jobs: INFORMATION:: new_job: <v2JobRecord(id=UUID('b3eb4c73-f8da-47cc-9e1e-9874d6ba07ef'))>
    [2024-04-19 19:03:45,344] INFO in jobs: Checking for remote build node for job
    [2024-04-19 19:03:45,890] INFO in jobs: Matching remote node: x9000c1s3b1n0, current jobs on node: 1
    ```

    Examine the information about the architecture of the artifact being used and the remote build nodes found.

## Clean up orphaned jobs

1. (`ncn-mw`) List the current IMS jobs to get the valid job IDs.

    ```bash

>     cray ims jobs list | jq '.[] | .id'

Reviewer: ```suggestion
    cray ims jobs list --format json | jq '.[] | .id'
```

    CONTAINER ID  IMAGE                                                            COMMAND     CREATED            STATUS            PORTS                 NAMES
    24438ea88859  localhost/ims-remote-020cb1fc-fbc4-4256-9a28-43b4527fc573:1.0.0              About an hour ago  Up About an hour           0.0.0.0:2022->22/tcp  ims-020cb1fc-fbc4-4256-9a28-43b4527fc573
    b4549f9b3608  localhost/ims-remote-b3eb4c73-f8da-47cc-9e1e-9874d6ba07ef:1.0.0              About an hour ago  Exited (1) 46 seconds ago  0.0.0.0:2023->22/tcp  ims-b3eb4c73-f8da-47cc-9e1e-9874d6ba07ef
    ca674a5de7cd  localhost/ims-remote-262af634-8cb7-4f8d-ab36-210ab6cc4409:1.0.0              About an hour ago  Up About an hour           0.0.0.0:2024->22/tcp  ims-262af634-8cb7-4f8d-ab36-210ab6cc4409
    40149c58272e  localhost/ims-remote-af620dcd-354f-4742-8d35-cc8fd7dea5a4:1.0.0              About an hour ago  Exited (1) 5 seconds ago   0.0.0.0:2025->22/tcp  ims-af620dcd-354f-4742-8d35-cc8fd7dea5a4
    3d7efd2de031  localhost/ims-remote-92319466-672a-4886-8bf6-3300096c74f8:1.0.0              52 minutes ago     Up 52 minutes              0.0.0.0:2026->22/tcp  ims-92319466-672a-4886-8bf6-3300096c74f8
    ```

>     Both the image and the name of the container have the ims JOB_ID incorporated.
> 
>     In the normal course of events, the IMS job will stop and remove all podman jobs on the remote node when the job
>     stops. There are times if the IMS job is stopped in an irregular way, the podman jobs may be stopped already -
>     check the 'STATUS' field for containers that have already exited.

Reviewer: ```suggestion
    Both the image and the name of the container have the IMS `JOB_ID` incorporated.

    In the normal course of events, the IMS job will stop and remove all podman jobs on the remote node when the job
    stops. There are times if the IMS job is stopped in an irregular way, the podman jobs may be stopped already -
    check the `STATUS` field for containers that have already exited.
```

1. (`cn#`) Stop any orphaned running containers.

>     Any containers that are still running, but is not in the IMS job list, will need to be stopped before they
>     can be removed.

Reviewer: ```suggestion
    If any containers are still running but are not in the IMS job list, then they will need to be stopped before they
    can be removed.
```

    ```

1. (`cn#`) Delete any dangling `podman` resources.

    To remove any additional resources that were attached to the removed containers and images,
    prune the system:

    ```bash
    podman volume prune -f
    ```

    There is usually no output from this command.

1. (`cn#`) Remove temporary directories.

    Jobs use temporary directories to transfer build artifacts back to Kubernetes. Each job will have
    a directory named `/tmp/ims_IMS_JOB_ID`. Delete any directories belonging to jobs that are no longer present.

    ```bash
    rm -rf /tmp/ims_OLD_IMS_JOB_ID
    ```

## Jobs fail due to lack of resources on the remote node

The most common cause for jobs failing to start on the remote node is a lack of resources on the remote node.
There are a couple of different ways these failures can manifest.

In either of the below cases, the solution is to either reduce the resource demands on the remote node through
running fewer concurrent jobs or jobs with smaller images, or to add storage to the remote node.

For more information, see [Adding storage to a remote build node](Configure_a_Remote_Build_Node.md#adding-storage-to-a-remote-build-node).

> ### IMS job fails in init

Reviewer: ```suggestion
### IMS job fails in `init`
```

> This usually happens when there isn't sufficient space to upload the remote job image to the remote node.

Reviewer: ```suggestion
This usually happens when there is insufficient space to upload the remote job image to the remote node.
```

    + SSHD_CONFIG_FILE=/etc/cray/ims/sshd_config
    Checking env vars
    + echo 'Checking env vars'
    + IMPORTED_VALS=('OAUTH_CONFIG_DIR' 'BUILD_ARCH' 'IMS_JOB_ID' 'IMAGE_ROOT_PARENT')
    + for item in "${IMPORTED_VALS[@]}"
    + [[ -z /etc/admin-client-auth ]]
    + for item in "${IMPORTED_VALS[@]}"
    + [[ -z x86_64 ]]
    + for item in "${IMPORTED_VALS[@]}"
    + [[ -z a818004a-4ee4-4670-a938-2a353960b803 ]]
    + for item in "${IMPORTED_VALS[@]}"

    ...

    Mounted /dev
    + '[' True = True ']'
    + echo 'ChrootDirectory /mnt/image/image-root/'
    + mkdir -p /root/.ssh
    + ssh-keygen -A
    ssh-keygen: generating new host keys: RSA DSA ECDSA ED25519 
    + echo 'SetEnv IMS_JOB_ID=a818004a-4ee4-4670-a938-2a353960b803 IMS_ARCH=x86_64 IMS_DKMS_ENABLED=True'
    + /usr/sbin/sshd -E /etc/cray/ims/sshd.log -f /etc/cray/ims/sshd_config
    + set +x
    off
    Waiting for signal file
    ```

### Look at the resources on the remote build node

There are two directories that are used by the remote builds and require the most space:

* `/tmp` - this is used to transfer files back and forth from the remote job to the system

> * `/var/lib/containers` - this is used by `podman` for images and container volume space

Reviewer: ```suggestion
* `/var/lib/containers` - this is used by podman for images and container volume space
```

> # Troubleshoot Issues with Remote Build Nodes

Reviewer: The main title of the page should match the filename (we have some exceptions to this in the docs repo, but mostly they are just from before we started following that standard)

Remote builds work by creating and running containerized jobs via `podman` on a remote node. There
are times that problems crop up with running these remote jobs.

> * [Troubleshoot Issues with Remote Build Nodes](#troubleshoot-issues-with-remote-build-nodes)

Reviewer: ```suggestion
```

* [Prerequisites](#prerequisites)
* [Jobs fail to launch on remote nodes](#jobs-fail-to-launch-on-remote-nodes)
* [Clean up orphaned jobs](#clean-up-orphaned-jobs)
* [Jobs fail on remote node](#jobs-fail-on-remote-node)

## Prerequisites

* Administrative access to a master or worker node on the K8S cluster
* Administrative access to the remote build node

## Jobs fail to launch on remote nodes

> 1. Verify that the [remote node](Configure_a_Remote_Build_Node.md#adding-and-removing-remote-build-nodes-to-ims) is defined in IMS
> 
> 1. Verify that [`ssh`](Configure_a_Remote_Build_Node.md#use-an-existing-compute-node#) keys are installed on remote node
> 
> 1. Verify that [`podman`](Configure_a_Remote_Build_Node.md#use-an-existing-compute-node) is installed on the remote build node.

Reviewer: ```suggestion
1. Verify that the remote node is defined in IMS.

    For more information, see [Adding and removing remote build nodes to IMS](Configure_a_Remote_Build_Node.md#adding-and-removing-remote-build-nodes-to-ims).

1. Verify that SSH keys are installed on the remote node.

    For more information, see [Use an existing compute node](Configure_a_Remote_Build_Node.md#use-an-existing-compute-node#).

1. Verify that `podman` is installed on the remote node.

    For more information, see [Use an existing compute node](Configure_a_Remote_Build_Node.md#use-an-existing-compute-node#).
```

1. Check the architecture of of the job and remote node.

>     (`ncn-mw#`) Check job architecture
> 
>     ```bash
>     cray ims jobs describe JOB_ID
>     ```
> 
>     (`cn#`) Check remote build node architecture
> 
>     ```bash
>     uname -a
>     ```

Reviewer: ```suggestion
    1. (`ncn-mw#`) Check job architecture.

        ```bash
        cray ims jobs describe JOB_ID
        ```

    1. (`cn#`) Check remote build node architecture.

        ```bash
        uname -a
        ```
```

> 1. (`ncn-mw#`) Check the IMS service logs for checks against the job. Example command:
> 

Reviewer: ```suggestion
1. (`ncn-mw#`) Check the IMS service logs for checks against the job.

    Example command:

```

    ```bash
    cray -n services logs cray-ims-69bc49b6bd-fm759
    ```

## Clean up orphaned jobs

> 1. (`cn#`) Check for the presence of leftover jobs on the remote build node

Reviewer: ```suggestion
1. (`cn#`) Check for the presence of leftover jobs on the remote build node.
```

    ```bash
    podman ps -a
    ```

> 1. (`cn#`) Stop any orphaned running containers

Reviewer: ```suggestion
1. (`cn#`) Stop any orphaned running containers.
```

    ```bash
    podman stop CONTAINER_ID
    ```

> 1. (`cn#`) Stop any orphaned running containers
> 
>     ```bash
>     podman rm CONTAINER_ID

Reviewer: ```suggestion
1. (`cn#`) Remove any orphaned running containers.

    ```bash
    podman rm CONTAINER_ID
```

    ```

> 1. (`cn#`) Check for the presence of additional container images on the system

Reviewer: ```suggestion
1. (`cn#`) Check for the presence of additional container images on the system.
```

    ```bash
    podman image ls
    ```

> 1. (`cn#`) Delete any images not being utilized by a valid running job

Reviewer: ```suggestion
1. (`cn#`) Delete any images not being utilized by a valid running job.
```

    ```bash
    podman rmi IMAGE_ID
    ```

> 1. (`cn#`) Delete any dangling `podman` resources

Reviewer: ```suggestion
1. (`cn#`) Delete any dangling `podman` resources.
```

    ```bash
    podman volume prune -f
    ```

> 1. (`cn#`) Jobs use temporary directories to transfer build artifacts back to `kubernetes`. Verify that `/tmp/ims_IMS_JOB_ID` for your valid running job is the only one present. Delete any unused directories.

Reviewer: ```suggestion
1. (`cn#`) Jobs use temporary directories to transfer build artifacts back to Kubernetes. Verify that `/tmp/ims_IMS_JOB_ID` for the valid running job is the only one present. Delete any unused directories.
```

    ```bash
    rm /tmp/ims_OLD_IMS_JOB_ID
    ```

## Jobs fail on remote node

> 1. Storage needs to be setup on the remote build node. Failure to do so will likely result in jobs failing. Ensure that the steps to setup [storage](Configure_a_Remote_Build_Node.md#adding-storage-to-a-remote-build-node) are completed.

Reviewer: ```suggestion
Storage needs to be setup on the remote build node. Failure to do so will likely result in jobs failing. For more information, see
[Adding storage to a remote build node](Configure_a_Remote_Build_Node.md#adding-storage-to-a-remote-build-node).
```
